File: packages/typescript/src/exporter/file.ts
/**
 * JSON-lines file exporter — for development, testing, air-gapped deploys.
 * Production should wrap with BoundedQueueExporter so disk I/O never
 * blocks the agent's hot path.
 */

import { appendFileSync, mkdirSync } from "node:fs";
import { dirname } from "node:path";

import { ErrorCode, logError } from "../internal/safelog.js";
import type { AgentCaptureSpan } from "../schema/span.js";

import type { SpanExporter } from "./base.js";

export class FileExporter implements SpanExporter {
  private closed = false;

  constructor(private readonly path: string) {
    mkdirSync(dirname(path), { recursive: true });
  }

  export(span: AgentCaptureSpan): void {
    if (this.closed) return;
    try {
      appendFileSync(this.path, JSON.stringify(span) + "\n", "utf-8");
    } catch (exc) {
      logError(ErrorCode.AC401, `FileExporter.export failed`, exc);
    }
  }

  async shutdown(_timeoutMs: number = 5000): Promise<void> {
    this.closed = true;
  }
}

File: packages/python/src/agent_capture/exporter/file.py
"""Synchronous JSON-lines file exporter.

For development, local testing, and air-gapped deployments. One JSON
object per line, UTF-8, append-mode. Thread-safe via a single lock.

This exporter blocks the caller on disk I/O. In production, wrap it in
:class:`agent_capture.exporter.queue.BoundedQueueExporter` so the agent's
hot path never waits on the filesystem.
"""

from __future__ import annotations

import os
import threading
from pathlib import Path

from agent_capture._internal.safelog import ErrorCode, log_error
from agent_capture.exporter.base import SpanExporter
from agent_capture.schema import Span


class FileExporter(SpanExporter):
    """Append finalized spans to a local JSON-lines file."""

    def __init__(self, path: str | os.PathLike[str]) -> None:
        self._path = Path(path)
        self._lock = threading.Lock()
        self._closed = False
        self._path.parent.mkdir(parents=True, exist_ok=True)

    @property
    def path(self) -> Path:
        return self._path

    def export(self, span: Span) -> None:
        if self._closed:
            return
        try:
            line = span.model_dump_json(exclude_none=False)
            with self._lock, self._path.open("a", encoding="utf-8") as fh:
                fh.write(line)
                fh.write("\n")
        except Exception as exc:
            log_error(ErrorCode.AC401, "FileExporter.export failed: %s", exc)

    def shutdown(self, timeout: float = 5.0) -> None:
        with self._lock:
            self._closed = True
